api/wikis/views.py

```diff
@@ -131,6 +141,17 @@ def get_serializer_class(self):
     def get_object(self):
         return self.get_wiki()
 
+    # overrides RetrieveUpdateDestroyAPIView
+    def perform_destroy(self, instance):
+        auth = get_user_auth(self.request)
+        wiki_page = self.get_object()
+        try:
+            wiki_page.node.delete_node_wiki(wiki_page, auth)
+        except NodeStateError:
+            raise MethodNotAllowed(method=self.request.method)
```

Comment:
Just a question - it took me awhile to figure out what would be throwing a NodeStateError until I saw you'd added it to the Registration model.  I am just wondering if this is the best way to do this.  How do we handle other objects that can belong to either a node or a registration in the API?

Was thinking about how we inherit from a generic ListCreateView, for example, to restrict what methods we could use. Maybe we could create something similar for these wiki views to inherit from to restrict the http methods if the wiki's node is a reg. Seems like we could consolidate this NodeStateError checking.

Feel free to push back.

Reply:
@pattisdr, this is an excellent note. I went back and forth on this as well, but I think it's most appropriate for validation to happen at the model level, not in the view (Chris Wisecarver beat the "skinny views, fat models" idea into me). Another reason I think it's preferable to have this validation at the model level is so that if someone attempts to call `delete_node_wiki` (or rename or edit) outside of the API (e.g. in a script), this error will be raised. 

However, I definitely agree that the `NodeStateError` is not easy to track down and I'm not sure how to improve that. Let me know if you have other thoughts on this. It's certainly an opinionated decision where I didn't consider all possibilities (such as like your mixin suggestion) 

From a (short) FD discussion with Steve:
```
Casey
Question of the day:
Wiki pages on registrations can’t be deleted. To prevent this from happening in the API, do we:
a) check is_registration in AbstractNode.delete_node_wiki and raise an exception that is handled by the API
b) override delete_node_wiki on the Registration model to raise an exception that is handled by the API
c) check wiki_page.node.is_registration in the WikiDetail API view and raise an exception

Casey
I would vote b

Casey
But we don’t seem to be doing that anywhere?

Steve
I second b
```

Reply:
@caseyrollins, I think whether or not we override `delete_node_wiki` and others on the Registration model is a separate issue. Whichever way you handle things API side, I think this was a good idea to further safeguard registrations.

The main reason I commented here, though, was after noticing the numerous NodeStateError checks across your views that occur well after the point in which we know the request will fail. For example, we've already validated the request data, which is unnecessary work if this is a disallowed request. As soon as the wiki is fetched, we have its node or registration, and therefore know if the request_method is acceptable.  

Similarly, there is no way to delete a registration through the API, we use a RetrieveUpdateAPIView on RegistrationDetail.  As soon as we send a Delete request to that endpoint it fails.  We don't have to wait to call the delete method on the registration to decide to throw a 405.

Reply:
Ah, I see what you mean. Not all of my `NodeStateError` checks are necessary (they'll never be hit, as you're saying). 

Here are the existing checks:
* [`NodeWikiSerializer.update()`](https://github.com/pattisdr/osf.io/pull/10/files#diff-dfa23dfe3da2849868aa007a7435fe5bR120): Definitely not needed as the request to update a registration wiki page would fail first [here](https://github.com/pattisdr/osf.io/pull/10/files#diff-dfa23dfe3da2849868aa007a7435fe5bR165). 
* [`WikiVersionSerializer.create()`](https://github.com/pattisdr/osf.io/pull/10/files#diff-dfa23dfe3da2849868aa007a7435fe5bR165): Arguably useful since both nodes and registrations share this serializer.
* [`WikiDetail.perform_destroy()`](https://github.com/pattisdr/osf.io/pull/10/files#diff-15e45414bd0cfad4288438347ef25273R150): Arguably useful since both node and registration wiki pages share this view.

However, I just now realized that the last two could be checked in [`WikiMixin.get_wiki()`](https://github.com/CenterForOpenScience/osf.io/blob/develop/api/wikis/views.py#L28) instead which will save us the wasted effort of validating request data, as you say. Let's go with that?

Reply:
That works great @caseyrollins , thanks